### ISSUE TITLE
Unexpected error behaviour: Parameterized generics cannot be used with class or instance checks

### ISSUE THREAD

**Issue by JohanSchott** [python/mypy#12155]
`mypy` reports success for
```python 
import numpy as np
x = "a"
assert isinstance(x, np.float64)
```
but fails for
```python
from numpy import float64
x = "a"
assert isinstance(x, float64)

```
with the error `fails.py:3: error: Parameterized generics cannot be used with class or instance checks`

I'm using `mypy` (0.931) and `numpy` (1.22.2).

Why do I get different results?

**Comment by JelleZijlstra** [comment]
Interesting. float64 is defined here: https://github.com/numpy/numpy/blob/51abd9693a78907f2ca7dfacee2017ef44fb9f49/numpy/__init__.pyi#L2973. It's obviously meant as a type alias, but apparently mypy sometimes interprets it as a variable assignment.

**Comment by lindycoder** [comment]
I have the same issue with Unions

version:
```
$  mypy --version
mypy 1.4.1 (compiled: yes)
```
Config:
```
[tool.mypy]
namespace_packages=true
explicit_package_bases=true
strict=true
show_error_codes=true
```

Code
```
from dataclasses import dataclass


@dataclass
class A:
    val_a: str

@dataclass
class B:
    val_b: str

ABUnion = A|B

obj = A(val_a="hey")

assert isinstance(obj, ABUnion)

```

Result:
```
$ mypy check.py 
check.py:16: error: Parameterized generics cannot be used with class or instance checks  [misc]
check.py:16: error: Argument 2 to "isinstance" has incompatible type "<typing special form>"; expected "_ClassInfo"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```


Code works fine


Thank you!

**Comment by tibbe** [comment]
Any status update?

Note that `isinstance` is defined as working on unions: "If *classinfo* is a tuple of type objects (or recursively, other such tuples) or a [Union Type](https://docs.python.org/3/library/stdtypes.html#types-union) of multiple types, return True if *object* is an instance of any of the types."

**Comment by brandon-neth** [comment]
I'm also wondering about the status on this.

**Comment by hunterC90** [comment]
Should this have the label `topic-pep-604`? I'm not 100% sure but it seems like it when I look at others open issues that have the label.

**Comment by msetina** [comment]
Any new development on mitigating this. 
I am using cryptography and their aliases are not disected properly and even get_args does not help to mark object as suitable.
Also a Union of child classes is not properly inspected when variable of object is of parent class but the object is of child class.

**Comment by gbhand** [comment]
@msetina cryptography seems to use [old-style unions](https://github.com/pyca/cryptography/blob/2f5b9502b097bed6c0a5db1a0cb176f4ec2c3d5c/src/cryptography/x509/base.py#L44-L53), which appear to still be unsupported https://github.com/python/mypy/issues/17680

@lindycoder's workaround has been successful as far as I've tried
```python
from numpy import float64
x = "a"
assert isinstance(x, float64)
```
yields `error: Parameterized generics cannot be used with class or instance checks  [misc]`

while this passes
```python
from typing import get_args
x = "a"
assert isinstance(x, get_args(float64))
```

with latest release
```
$ mypy --version
mypy 1.17.0 (compiled: yes)
```

Relating to cryptography the workaround also seems to do the trick for the general case, can you give a more specific example of the inheritance-related failure?
```python
from cryptography.hazmat.primitives.hashes import SHA256
from cryptography.x509.base import _AllowedHashTypes

assert isinstance(SHA256, _AllowedHashTypes) # fails with [misc, arg-type]
assert isinstance(SHA256, get_args(_AllowedHashTypes)) # passes
```
